### PR TITLE
Add extracting(String, InstanceOfAssertFactory) to AbstractObjectAssert

### PR DESCRIPTION
**Pull request by scordio**
#### Check List:
* Fixes N/A
* Unit tests : YES
* Javadoc with a code example (API only) : YES

### CODE REVIEW

File: src/main/java/org/assertj/core/api/AbstractObjectAssert.java [hunk 1]
```diff
@@ -663,6 +662,55 @@ public SELF hasFieldOrPropertyWithValue(String name, Object value) {
...
+   * for, if the field is not accessible (i.e. does not exist) an {@link IntrospectionError} is thrown.
+   *
+   * @param <ASSERT>        the type of the resulting {@code Assert}
+   * @param propertyOrField the property/field to extract from the initial object under test
+   * @param assertFactory   the factory which verifies the type and creates the new {@code Assert}
+   * @return a new narrowed {@link Assert} instance whose object under test is the extracted property/field value
+   * @throws NullPointerException if the given factory is {@code null}
+   * @throws IntrospectionError if one of the given name does not match a field or property
+   *
+   * @since FIXME 3.13.3 or 3.14.0?
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractObjectAssert.java]: 3.14.0

File: src/test/java/org/assertj/core/api/object/ObjectAssert_extracting_with_String_and_InstanceOfAssertFactory_Test.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    // WHEN
+    AbstractStringAssert<?> result = assertThat(luke).extracting("name.first", STRING);
+    // THEN
+    result.startsWith("Lu");
+  }
+
+  @Test
+  void should_use_property_field_name_as_description_when_extracting_single_property() {
+    // WHEN
+    Throwable thrown = catchThrowable(() -> assertThat(luke).extracting("name.first", STRING)
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/object/ObjectAssert_extracting_with_String_and_InstanceOfAssertFactory_Test.java]: use [`expectAssertionError`](https://github.com/joel-costigliola/assertj-core/blob/a58805bbbeb6c5fbbd7b0277d1d66518d7400c65/src/test/java/org/assertj/core/util/AssertionsUtil.java#L24) utility method to directly get the `AssertionError`, same comment for the test below

File: src/test/java/org/assertj/core/api/object/ObjectAssert_extracting_with_String_and_InstanceOfAssertFactory_Test.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+import org.assertj.core.api.ObjectAssert;
+import org.assertj.core.test.Employee;
+import org.assertj.core.test.Name;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+/**
+ * Tests for <code>{@link ObjectAssert#extracting(String, InstanceOfAssertFactory)}</code>.
+ */
+class ObjectAssert_extracting_with_String_and_InstanceOfAssertFactory_Test {
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/object/ObjectAssert_extracting_with_String_and_InstanceOfAssertFactory_Test.java]: let's add a test for soft assertions (`SoftAssertionsTest`) and assumptions (`Object_special_assertion_methods_in_assumptions_Test`) to make sure they work fine with the new method (it should be the case but it's better to be sure with a test).

**Comment by joel-costigliola** [comment]
@scordio PR reviewed, thanks.

**Comment by scordio** [comment]
@joel-costigliola PR ready!

**Comment by joel-costigliola** [comment]
Integrated thanks @scordio ! I have just a few tests notably one for soft assertions that extracting assertion errors are properly collected (this test should have been added way back but was not)

https://github.com/joel-costigliola/assertj-core/blob/0203433d7b028eab5faa187bdaa084f6613b0885/src/test/java/org/assertj/core/api/SoftAssertionsTest.java#L653

**Comment by scordio** [comment]
Thanks! I missed that :man_facepalming:

**Comment by joel-costigliola** [comment]
Not your fault, there was no tests for extracting on object !

**Pull request closed without merging**